gateway-sp-comms/src/single_sp.rs

```diff
@@ -236,55 +245,184 @@ impl SingleSp {
-                        .map_err(Into::into)
-                        .map(|response| (response, data))
-                })?;
+                    // Interpret the data as UTF8.
+                    let device =
+                        str::from_utf8(&data[..device_len]).map_err(|_| {
+                            SpCommunicationError::TlvPagination {
+                                reason: "non-UTF8 device",
+                            }
+                        })?;
+                    let description = str::from_utf8(&data[device_len..])
+                        .map_err(|_| SpCommunicationError::TlvPagination {
+                            reason: "non-UTF8 description",
+                        })?;
+
+                    Ok(Some(SpDevice {
+                        component: header.component,
+                        device: device.to_string(),
+                        description: description.to_string(),
+                        capabilities: header.capabilities,
+                        presence: header.presence,
+                    }))
+                }
+                _ => {
+                    info!(
+                        self.log,
+                        "skipping unknown device description tag {tag:?}"
+                    );
+                    Ok(None)
+                }
+            },
+        )
+        .await
+        .map(|devices| SpInventory { devices })
+    }
+
+    /// Request the detailed status / measurements of a particular component
+    /// accessible to the SP.
+    pub async fn component_details(
+        &self,
+        component: SpComponent,
+    ) -> Result<SpComponentDetails> {
+        self.get_paginated_tlv_data(
+            |offset| MgsRequest::ComponentDetails { component, offset },
+            |response| response.expect_component_details(),
+            |tag, value| match tag {
+                PortStatus::TAG => {
+                    let (result, leftover) = gateway_messages::deserialize::<
+                        Result<PortStatus, PortStatusError>,
+                    >(value)
+                    .map_err(|err| SpCommunicationError::TlvDeserialize {
+                        tag,
+                        err,
+                    })?;
+
+                    if !leftover.is_empty() {
+                        info!(
+                            self.log,
+                            "ignoring unexpected data in PortStatus TLV entry"
+                        );
+                    }
+
+                    Ok(Some(ComponentDetails::PortStatus(result)))
+                }
+                _ => {
+                    info!(
+                        self.log,
+                        "skipping unknown component details tag {tag:?}"
+                    );
+                    Ok(None)
+                }
+            },
+        )
+        .await
+        .map(|entries| SpComponentDetails { entries })
+    }
+
+    async fn get_paginated_tlv_data<T, F, G, H>(
```

Comment:
It took me a little bit of time to get a grasp of what was going on here, but it seems like a decent abstraction given the two different caller so far. I do wonder if there's a cleaner, mechanism so the logic flows linearly at the call site rather than in closures, but this definitely works. 

I don't think you need to change anything in this commit necessarily, or even at all, but I'd like to spitball with you if you'd indulge me.

It may be possible to make this abstraction flow better by creating a `TlvParser `class (possibly with typestates) and then chaining methods that take closures rather than have the closures all passed into one method function up front. It would look something like this I believe for `component_details`:

```rust
   TlvParser::new()
              .request(|offset| MgsRequest::ComponentDetails {
                  component,
                  offset,
              })
              .await?
              .expect_response(|response| response.expect_component_details())?
              .decode(|tag, value| ())
      }
```

Reply:
Maybe change the name of the class to `TlvRpc` to reflect the role of the remote call.

Reply:
Continuing the spitball - what about getting rid of the closures entirely via a trait with two implementors? Untested but roughly

```rust
trait TlvRpc {
    type Item;

    fn request(&self, offset: u32) -> MgsRequest;
    fn parse_response(&self, response: SpResponse) -> Result<TlvPage>;
    fn decode(&self, tag: Tag, value: &[u8]) -> Result<Item>;
}

async fn run_tlvc_rpc<T: TlvRpc>(runner: T) -> Result<Vec<T::Item>> { .. }
```

Reply:
This is much better! I knew you'd fix my shenanigans :)